alibi/explainers/partial_dependence.py

```diff
@@ -0,0 +1,1294 @@
+
+        return response_method, method, kind
+
+    def _partial_dependence(self,
+                            estimator: BaseEstimator,
+                            X: np.ndarray,
+                            features: Union[int, Tuple[int, int]],
+                            response_method: Literal['auto', 'predict_proba', 'decision_function'] = 'auto',
+                            method: Literal['auto', 'recursion', 'brute'] = 'auto',
+                            kind: Literal['average', 'individual', 'both'] = 'average',
+                            percentiles: Tuple[float, float] = (0., 1.),
+                            grid_resolution: int = 100,
+                            grid_points: Optional[Dict[int, np.ndarray]] = None
+                            ) -> Dict[str, np.ndarray]:
+        """
+        Computes partial dependence for a feature or a pair of features.
+
+        Parameters
+        ----------
+        estimator, X, features, response_method, method, kind, percentiles, grid_resolution, grid_points
+            See :py:meth:`alibi.explainers.partial_dependence.PartialDependence.explain` method.
+
+        Returns
+        -------
+        A dictionary containing the feature(s) values, feature(s) deciles, average and/or individual values
+        (i.e. partial dependence or individual conditional expectation) of the give (pair) of feature(s))
+        """
+        if isinstance(features, numbers.Integral):
+            features = (features, )
+
+        if grid_points is None:
+            grid_points = {}
+
+        deciles, values, features_indices = [], [], [],
+        for f in features:  # type: ignore[union-attr]
+            # Note that we are using the _safe_indexing procedure implement in sklearn to retrieve the column.
+            # This is because the _safe_indexing supports many data types such as sparse matrix representation,
+            # pandas dataframes etc. In case we would like to enhance alibi in the future to support other
+            # data types than np.ndarrays, this can be useful.
+            f_indices = np.asarray(_get_column_indices(X, f), dtype=np.int32, order='C').ravel()
+            X_f = _safe_indexing(X, f_indices, axis=1)
+
+            # get deciles for the current feature if the feature is numerical
+            deciles_f = get_quantiles(X_f, num_quantiles=11) if self._is_numerical(f) else None
+
+            if f not in grid_points:
+                # construct grid for feature f. Note that for categorical features we pass the
+                # grid resolution to be infinity because otherwise we risk to apply `linspace` to
+                # categorical values, which does not make sense.
+                _, values_f = _grid_from_X(X_f,
+                                           percentiles=percentiles,
+                                           grid_resolution=grid_resolution if self._is_numerical(f) else np.inf)
```

Comment:
yes. This is the logic that allows us to control the grid resolution for numerical features and also treat correctly the categorical feature case. For example, consider that we want to compute the 2-way pdp for a numerical and a categorical feature. Now lets say that the numerical feature has 1000 feature values and the categorical one has 15 feature values. If we were to use directly the sklearn API and set the grid points to something less than 15, we would have ran into some problems. This is because the backend would have generated a grid point for the categorical values too. Instead of leaving them untouched (i.e. [0, 1, 2, ... , 14]) would have generated some float numbers (e.g. [0, 2.5, 7.5, 12.5]), which is non-sense.